web/yo/app/scripts/services/drugmaputils.js

```diff
@@ -0,0 +1,176 @@
+                    break;
+                case 'tumor':
+                    break;
+                case 'treatment':
+                    //Functions here should be optimized.
+                    _.forEach(drugArray, function(drug){
+                        var chainDrug = $q.when();
+                        var mapPath = '';
+                        chainDrug = chainDrug.then(function(){
+                            mapPath = getMapPath(drug, geneName, mutationUuid, cancerTypeUuid, name_uuid);
+                            firebaseConnector.mapOnce(mapPath + '/status').then(function (result){
+                                if(result === 'latest'){
+                                    firebaseConnector.removeMap(mapPath).then(function () {
+                                        firebaseConnector.ref('Map/' + drug + '/' + geneName + '/' + mutationUuid + '/cancerTypes').once('value', function (doc) {
+                                            if (!doc.val()) {
+                                                mapPath = drug + '/' + geneName + '/' + mutationUuid + '/mutationName';
+                                                firebaseConnector.removeMap(mapPath);
+                                            }
+                                        })
+                                    })
+                                }
+                            })
+                        })
+                    });
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        function changeTherapyNameInMapWithValidatingStatus(dataType, geneName, mutationUuid, mutationName, cancerTypeUuid, therapyUuid, drugArray, content){
+            switch (dataType){
+                case 'mutation':
+                    break;
+                case 'tumor':
+                    break;
+                case 'treatment':
+                    _.forEach(drugArray, function(drug){
+                        var mapPath = '';
+                        var chainDrug = $q.when();
+                        chainDrug = chainDrug.then(function(){
+                            mapPath = drug + '/' + geneName + '/' + mutationUuid + '/cancerTypes/' + cancerTypeUuid + '/' + therapyUuid;
+                            firebaseConnector.mapOnce(mapPath + '/status').then(function (result){
+                                if(result === 'latest'){
+                                    mapPath = drug + '/' + geneName + '/' + mutationUuid + '/cancerTypes/' + cancerTypeUuid + '/' + therapyUuid + '/name';
+                                    firebaseConnector.setMap(mapPath, content);
+                                }
+                            })
+                        })
+                    });
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        function getMapPath(drug, geneName, mutationUuid, cancerTypeUuid, therapyUuid){
```

Comment:
This needs to be done more thoughtfully. The purpose of this function is getting the path to the therapyInMap.
Also please take a look all other places where you generate the path. These should all be done in a Util. Maybe FirebasePathUtils. For instance, in function `changeTherapyNameInMapWithValidatingStatus`